### PR TITLE
[DataPipe] implementing demux()

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64215
* __->__ #63650
* #63649

Differential Revision: [D30493944](https://our.internmc.facebook.com/intern/diff/D30493944)

cc @VitalyFedyunin @ejguan

### CODE REVIEW

File: torch/utils/data/datapipes/iter/combining.py [hunk 1]
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
...
-    # Placeholder implementation
-    def __new__(cls, datapipe, instances, classifier_fn):
-        result = []
-        buffer = list(datapipe)
+    def _find_next(self, instance_id: int) -> T_co:
+        while True:
+            value = self.main_datapipe.__next__()
+            classification = self.classifier_fn(value)
+            if classification > self.num_instances:
+                raise ValueError(f"Output of the classification fn should be between 0 and {self.num_instances - 1}. " +
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/combining.py:180]: Add test

File: torch/utils/data/datapipes/iter/combining.py [hunk 2]
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
...
+        as requested by the child DataPipes.
+    """
+
+    def __init__(self, datapipe: IterDataPipe[T_co], num_instances: int,
+                 classifier_fn: Callable[[T_co], int], buffer_size: int):
+        self.main_datapipe = iter(datapipe)
+        self.num_instances = num_instances
+        self.buffer_size = buffer_size
+        self.child_buffers: List[Deque[T_co]] = [deque() for _ in range(num_instances)]
+        self.current_buffer_total = 0
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/combining.py:172]: nit: s/current_buffer_total/total_buffers_size/g

File: torch/utils/data/datapipes/iter/combining.py [hunk 3]
```diff
@@ -137,24 +135,70 @@ def __iter__(self):
...
-    def __init__(self, datapipe: IterDataPipe[T_co], classifier_fn: Callable[[T_co], int]):
-        self.datapipe = datapipe
+        Container to hold instance-specific information on behalf of DemultiplexerIterDataPipe. It tracks
+        the state of its child DataPipes, maintains the buffer, classifies and yields the next correct value
+        as requested by the child DataPipes.
+    """
+
+    def __init__(self, datapipe: IterDataPipe[T_co], num_instances: int,
+                 classifier_fn: Callable[[T_co], int], buffer_size: int):
+        self.main_datapipe = iter(datapipe)
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/combining.py:168]: drop `iter` please

**Review by VitalyFedyunin — COMMENTED** [review]
Couple changes required

File: torch/utils/data/datapipes/iter/combining.py (hunk 2 above)
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combining.py:172]: I renamed the variables to
```
self.max_buffer_size = buffer_size
self.current_buffer_usage = 0
```

File: torch/utils/data/datapipes/iter/combining.py (hunk 3 above)
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combining.py:168]: As discussed in the other PR, I need `iter` for the subsequent `__next__` call.

File: torch/utils/data/datapipes/iter/combining.py (hunk 1 above)
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combining.py:180]: Added.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Review by ejguan — COMMENTED** [review]
A few comments similar to `Fork`

**Comment by ejguan** [comment]
Can you please also add `Demux` and `Mux` to https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/iter/__init__.py to expose these APIs for users?

**Comment by ejguan** [comment]
Forward feedback from vision domain for `demux`. Can we provide a flag of `drop_none`? When the result of `classification_fn` is `None`, `Demux` can ignore the data without putting it into buffer.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Review by ejguan — APPROVED** [review]
LGTM

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30493944).

**Pull request closed without merging**